File: app/src/main/java/co/ninetynine/android/exercise/adapters/SectionsListAdapter.java
package co.ninetynine.android.exercise.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;

import co.ninetynine.android.exercise.R;
import co.ninetynine.android.exercise.model.Section;
import co.ninetynine.android.exercise.viewholders.SectionViewHolder;

public class SectionsListAdapter extends RecyclerView.Adapter<SectionViewHolder> {

    private ArrayList<Section> sections;

    public SectionsListAdapter(ArrayList<Section> sections) {
        this.sections = sections;
    }

    @NonNull
    @Override
    public SectionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new SectionViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_section, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull SectionViewHolder holder, int position) {
        holder.bind(sections.get(position));
    }

    @Override
    public int getItemCount() {
        return sections.size();
    }
}

File: app/src/main/java/co/ninetynine/android/exercise/adapters/SectionsListAdapter.kt
package co.ninetynine.android.exercise.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import co.ninetynine.android.exercise.databinding.LayoutSectionBinding
import co.ninetynine.android.exercise.listeners.ViewInteractionListener
import co.ninetynine.android.exercise.model.Section
import co.ninetynine.android.exercise.viewholders.SectionViewHolder

class SectionsListAdapter(
        private val sections: ArrayList<Section>
) : RecyclerView.Adapter<SectionViewHolder>() {

    private val displayedSetions: MutableList<Section> = sections.filter { it.isVisible }.toMutableList()

    private val listener = object : ViewInteractionListener {
        override fun onInteract() {
            displayedSetions.clear()
            displayedSetions.addAll(sections.filter { it.isVisible })
            notifyDataSetChanged()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            SectionViewHolder(
                    LayoutSectionBinding.inflate(
                            LayoutInflater.from(parent.context),
                            parent,
                            false
                    ),
                    listener
            )

    override fun onBindViewHolder(holder: SectionViewHolder, position: Int) {
        holder.bind(displayedSetions[position])
    }

    override fun getItemCount() = displayedSetions.size
}